#Taller 2


*   Roberto José Ávila Moreno
*   Salim Abou-Ammar Garcia



In [ ]:
# Import packages.
import cvxpy as cp
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog

##Punto 1

$$ Max. Ganancia = 3A+5B$$
$$Suj. Produccion:$$
$$1A+1B≤500$$
$$2A+4B≤1000$$
$$1A+2B≤2000$$
$$A+0(B)≥100$$
$$0(A)+B≥150$$

In [ ]:
#Funcion Objetivo
C2 = np.array([3,5])


#Restricciones de Desigualdad
A2 = np.array([[1,1],
               [2,4],
                [1,2],
               [-1,0],
               [0,-1]])

B2 = np.array([500,1000,2000,-100,-150])


#Solución
solucion = linprog(-C2, A_ub=A2, b_ub=B2, method='revised simplex')
solucion

<ipython-input-33-019a4ec91441>:16: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solucion = linprog(-C2, A_ub=A2, b_ub=B2, method='revised simplex')


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -1350.0
       x: [ 2.000e+02  1.500e+02]
     nit: 2

Ganancias Maximas Totales, interpretadas en valor absoluto.

In [ ]:
-solucion.fun

1350.0

In [ ]:
Resultados_1 = pd.DataFrame(solucion.x)
Resultados_1 = Resultados_1.rename(columns={0:'Cantidades'})
Resultados_1 = Resultados_1.set_axis(['A', 'B'], axis=0)
Resultados_1

Cantidades
A       200.0
B       150.0

La asignación optima que permite maximizar la ganancia de la empresa, teniendo en cuenta las restricciones, es de 200 cantidades de A y 150 cantidades de B. Para este caso con dicha asignación obtenemos una ganancia de $1,350

##Punto 2

$$Max. Retorno = 0.10𝑥1+0.15𝑥2+0.16𝑥3+0.08𝑥4$$

$$0.5 𝑥1 + 0.3 𝑥2 + 0.25 𝑥3 + 0.6 𝑥4 ≥ 0.35 (valor)$$
$$0.3 𝑥1 + 0.1 𝑥2 + 0.4 𝑥3 + 0.2 𝑥4 ≥ 0.30 (tamaño)$$
$$0.2 𝑥1 + 0.6 𝑥2 + 0.35 𝑥3 + 0.2 𝑥4 ≥ 0.15 (liquidez)$$
$$𝑥1 + 𝑥2 + 𝑥3 + 𝑥4 = $100$$
$$𝑥1, 𝑥2, 𝑥3, 𝑥4 ≥ 0$$

###Metodo Scipy

ub = desigualdad /
eq = igualdad

In [ ]:
#Funcion Objetivo
Obj = np.array([0.10,0.15,0.16,0.08])


#Restricciones de Desigualdad
R1 = np.array([[0.5,0.3,0.25,0.6],
               [0.3,0.1,0.4,0.2],
               [0.2,0.6,0.35,0.2]
               ])


#Restricciones de Igualdad
R2 = np.array([[1,1,1,1]
               ])

# Linprog Arroja resultados no negativos, por lo que no hay que poner esta restriccion

#Resultados
Res1 = np.array([35,30,15]) #En la misma unidad de medida
Res2 = np.array([100])


#Solución
solucion = linprog(-Obj, A_ub=-R1, b_ub=-Res1, A_eq=R2, b_eq=Res2, method='revised simplex') # -R1 decir al menos (mayor o igual) (limite inferior) 
solucion

<ipython-input-37-20c4e8159bac>:24: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solucion = linprog(-Obj, A_ub=-R1, b_ub=-Res1, A_eq=R2, b_eq=Res2, method='revised simplex') # -R1 decir al menos (mayor o igual) (limite inferior)


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -13.736842105263158
       x: [ 0.000e+00  1.579e+01  5.789e+01  2.632e+01]
     nit: 7

Retorno Esperado Maximo, Interpretado en valor absoluto (En pesos)

In [ ]:
-solucion.fun

13.736842105263158

In [ ]:
Resultados_2 = pd.DataFrame(solucion.x)
Resultados_2 = Resultados_2.rename(columns={0:'Pesos'})
Resultados_2 = Resultados_2.set_axis(['W Fondo 1', 'W Fondo 2','W Fondo 3','W Fondo 4'], axis=0)
Resultados_2

Pesos
W Fondo 1   0.000000
W Fondo 2  15.789474
W Fondo 3  57.894737
W Fondo 4  26.315789

###Metodo CVXPY

In [ ]:
# Define and solve the CVXPY problem.
n = 4 #numero de incognitas
x = cp.Variable(n) #vector x con el n de incognitas
objective = cp.Minimize(-Obj@x) #Lo que hare con la funcion objetivo, que en este caso es maximizar (min negativo)
contraints = [R1@x >= Res1,cp.sum(x)==100,x>=0] #Restricciones
prob = cp.Problem(objective, contraints)

#Result
resultado = prob.solve()

#Valores optimos de x
x_valores = x.value
x_solucion = pd.DataFrame(x.value)
print(round(x_solucion,2))

       0
0   0.00
1  15.79
2  57.89
3  26.32


Como fue comprobado con los 2 metodos, la asginacion ideal de pesos a cada fondo, para lograr maximizar el retorno y teniendo encuenta las restricciones, es de 0 para el fondo 1, de 15.79 para el fondo 2, de 57.89 para el fondo 3 y de 26.32 para el fondo 4.

##Punto 3

$$Max.z6$$

$$Suj.Producción$$
$$x_1+y_1-z_1=150$$
$$x_2+y_2-1.01x_1+1.003z_1-z_2=150$$
$$x_3+y_3-1.01x_2+1.003z_2-z_3=-200$$
$$x_4-1.02y_1-1.01x_3+1.003z_3-z_4=200$$
$$x_5-1.02y_2-1.01x_4+1.003z_4-z_5=-50$$
$$-1.02y_3-1.01x_5+1.003z_5-z_6=-300$$

###Metodo Scipy

In [ ]:
#Funcion objetivo

C3 = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,1])

##Restricciones de igualdad

A2 = np.array([[1,0,0,0,0,1,0,0,-1,0,0,0,0,0],
               [-1.01,1,0,0,0,0,1,0,1.003,-1,0,0,0,0],
               [0,-1.01,1,0,0,0,0,1,0,1.003,-1,0,0,0],
               [0,0,-1.01,1,0,-1.02,0,0,0,0,1.003,-1,0,0],
               [0,0,0,-1.01,1,0,-1.02,0,0,0,0,1.003,-1,0],
               [0,0,0,0,-1.01,0,0,-1.02,0,0,0,0,1.003,-1]
               ])


B2 = np.array([150,100,-200,200,-50,-300])

#Solución
solucion = linprog(-C3,A_eq=A2, b_eq=B2, method='revised simplex')
solucion

<ipython-input-13-d3c50de919aa>:19: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  solucion = linprog(-C3,A_eq=A2, b_eq=B2, method='revised simplex')


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -92.49694915254233
       x: [ 0.000e+00  5.098e+01 ...  0.000e+00  9.250e+01]
     nit: 6

In [ ]:
-solucion.fun

-92.49694915254233

In [ ]:
pd.DataFrame(solucion.x)

0
0     0.000000
1    50.980392
2     0.000000
3     0.000000
4     0.000000
5   150.000000
6    49.019608
7   203.434364
8     0.000000
9     0.000000
10  351.944167
11    0.000000
12    0.000000
13   92.496949

###Metodo CVXPY

In [ ]:
# Define and solve the CVXPY problem.
n = 14 #numero de incognitas
x = cp.Variable(n) #vector x con el n de incognitas
objective = cp.Minimize(-C3@x) #Lo que hare con la funcion objetivo, que en este caso es maximizar (min negativo)
contraints = [A2@x == B2, x>=0] #Restricciones X son cantidades de dinero, si lo x son negativos tengo que especificar eso.
prob = cp.Problem(objective, contraints)

#Result
resultado = prob.solve(solver=cp.GLPK_MI)

#Valores optimos de x
x_valores = x.value
x_solucion = pd.DataFrame(x.value)
print(round(x_solucion,2))

         0
0    -0.00
1    50.98
2    -0.00
3    -0.00
4    -0.00
5   150.00
6    49.02
7   203.43
8    -0.00
9    -0.00
10  351.94
11   -0.00
12   -0.00
13   92.50


Como se evidencia con la ayuda de los dos metodos, se necesita 92.50 para maximizar el saldo de los fondos en el mes 6, con respecto a las restricciones requeridasd cada mes.

##Punto 4

$$Max. Rentabilidad$$
$$Suj: diff (ent + salidas)$$
$$semana 1 = x + y = 150$$

$$semana 2 = (-y*0.001) + x2 + y2 = 100$$

$$semana 3 =  (-y2*0.001) + x3 + y3 = -50$$

$$semana 4 =  (-y3*0.001) + x4 + y4 = 200$$

In [ ]:
C4 = np.array([0,0,0,0,0,0,0,1])
ren4 = (1+0.0001)**7
A4 = np.array([[1,0,0,0,-1,0,0,0],
               [0,1,0,0,ren4,-1,0,0],
               [0,0,1,0,0,ren4,-1,0],
               [0,0,0,1,0,0,ren4,-1,]])
b4 = np.array([[150,-100,-50,200]])
r4 = linprog(-C4, A_ub=A4, b_ub=-b4, method='revised simplex')
r4

<ipython-input-56-58eebd34be23>:8: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  r4 = linprog(-C4, A_ub=A4, b_ub=-b4, method='revised simplex')


 message: The problem is unbounded, as the simplex algorithm found a basic feasible solution from which there is a direction with negative reduced cost in which all decision variables increase.
 success: False
  status: 3
     fun: -200.14021366142958
       x: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00  1.500e+02
            5.011e+01  1.401e-01  2.001e+02]
     nit: 3

In [ ]:
-r4.fun

200.14021366142958

In [ ]:
pd.DataFrame(r4.x).T

0    1    2    3      4          5         6           7
0  0.0  0.0  0.0  0.0  150.0  50.105032  0.140116  200.140214

Para maximizar la rentabilidad total de los saldos en las cuentas teniendo como referencia las restricciones de flujos de efectivo se distribuye asi: 150 para el mes 4, 50.105032 para el mes 5, 0.140616 para el mes 6 y 200.140214 para el mes 7

##Punto 5

$$Max. Ganancias = 5𝑥1+8𝑥2+10𝑥3$$


$$𝑥1+𝑥2+𝑥3=10.000(Presupuesto)$$
$$𝑥1+0+0≥2.000$$
$$0+𝑥2+0≥2.000$$
$$𝑥1+𝑥2+𝑥3=1.000(Horas. de. Trabajo)$$
$$0+𝑥2+0≥300%$$
$$0+0+𝑥3≥250$$


X1=ventas --
X2=marketing --
X3=producción

In [ ]:
#Agregar un valor en el vector de desigualdad y en la matriz de desigualdades
#Funcion Objetivo
Obj = np.array([5,8,10,0,0,0])


#Restricciones de Desigualdad (mapear, identificar la variable)
R1 = np.array([[1,0,0,0,0,0],
               [0,1,0,0,0,0],
               [0,0,0,0,1,0],
               [0,0,0,0,0,1],
               #[#Una linea]
               ])

#Restricciones de Igualdad
R2 = np.array([[0,0,0,1,1,1],
               [1,1,1,0,0,0]
               ])

# Linprog Arroja resultados no negativos, por lo que no hay que poner esta restriccion

#Resultados
Res1 = np.array([10000*0.2,10000*0.2,1000*0.3,1000*0.25]) #En la misma unidad de medida
Res2 = np.array([1000,10000])


#Solución
solucion = linprog(-Obj, A_ub=-R1, b_ub=-Res1, A_eq=R2, b_eq=Res2, method='simplex') # -R1 decir al menos (mayor o igual) (limite inferior). Trabajar todo  
pd.DataFrame(solucion.x)

Ganancia Maxima, Interpretada en valor absoluto (En pesos)

In [ ]:
-solucion.fun

86000.0

In [ ]:
Resultados_5 = pd.DataFrame(solucion.x)
Resultados_5 = Resultados_5.rename(columns={0:'Resultados'})
Resultados_5 = Resultados_5.set_axis(['Presupuesto Ventas', 'Presupuesto Marketing','Presupuesto Producción','Horas de Trabajo Ventas','Horas de Trabajo Marketing','Horas de Trabajo Prudcción'], axis=0)
Resultados_5

Resultados
Presupuesto Ventas              2000.0
Presupuesto Marketing           2000.0
Presupuesto Producción          6000.0
Horas de Trabajo Ventas          450.0
Horas de Trabajo Marketing       300.0
Horas de Trabajo Prudcción       250.0

Para obtener una ganancia maxima para la empresa, esta debe hacer las siguientes asignaciones, para el departamento de ventas debera asignar 2,000 del presupuesto total y 450 horas de trabajo, para el departamento de marketing debera asignar 2,000 del presupuesto total y 300 horas de trabajo y por ultimo para el departamento de producción debera asignar 6,000 del presupuesto total y 250 horas de trabajo. Con esto la empresa lograria obtener su maxima ganancia que es de $86,000.

#Punto 6

$$Min. CostoTotal = Cji * Xji + F_j * Y_i $$


In [ ]:
#Para hacerlo con CVXPY

#Variables booleanas (valores de 0 y 1)
#Cantidades no negativas

#Funcion Objetivo (Mini)
Obj= np.array([[4,5,6,8,10],
              [6,4,3,5,8],
              [9,7,4,3,4]]) #Gastos de Distribucion
Obj2= np.array([1000,1000,1000]) #Gastos de Funcionamiento

R3= np.array([500,500,500]) #Cantidades Maximas de produccion por fabrica
R4= np.array([80,270,250,160,180]) #Demandas Requeridas


x = cp.Variable((3,5), nonneg=True) #Cantidades totales a distribuir. Matriz Obj. Nonneg para no tomar negativos
y = cp.Variable(3, boolean=True) #Restriccion de 1000 de funcionamiento para cada planta
objective = cp.Minimize((cp.sum(cp.multiply(Obj,x))+(cp.sum(cp.multiply(Obj2,y))))) #Lo que hare con la funcion objetivo, que en este caso es minimizar
contraints = [cp.sum(x[0,:])<= R3[0],cp.sum(x[1,:])<= R3[1],cp.sum(x[2,:])<= R3[2],cp.sum(x[:,0])==R4[0],
               cp.sum(x[:,1])==R4[1],cp.sum(x[:,2])==R4[2],cp.sum(x[:,3])==R4[3],
               cp.sum(x[:,4])==R4[4],x[0,:]==0] #Restricciones

prob = cp.Problem(objective, contraints)

#Result
resultado = prob.solve()

#Valores optimos de x
x_valores = x.value
x_solucion = pd.DataFrame(x.value)
print(round(x_solucion,2))

      0      1      2      3      4
0  -0.0   -0.0   -0.0   -0.0   -0.0
1  80.0  270.0  150.0   -0.0   -0.0
2  -0.0   -0.0  100.0  160.0  180.0


In [ ]:
Resultados_6 = pd.DataFrame(x.value)
Resultados_6 = Resultados_6.set_axis(['Planta 1', 'Planta 2','Planta 3'], axis=0)
Resultados_6 = Resultados_6.set_axis(['Punto de Demanda 1', 'Punto de Demanda 2','Punto de Demanda 3','Punto de Demanda 4','Punto de Demanda 5'], axis=1)
Resultados_6

Punto de Demanda 1  Punto de Demanda 2  Punto de Demanda 3  \
Planta 1                -0.0                -0.0                -0.0   
Planta 2                80.0               270.0               150.0   
Planta 3                -0.0                -0.0               100.0   

          Punto de Demanda 4  Punto de Demanda 5  
Planta 1                -0.0                -0.0  
Planta 2                -0.0                -0.0  
Planta 3               160.0               180.0

Para este caso cerramos la planta numero 1, por lo que nos queda que para minimizar los costos se va a distribuir la mercancia de las plantas a los puntos de demanda de la siguiente manera: la planta 2 cubrira toda la demanda del punto 1 (Cantidad total demanda y abastecida: 80) y punto 2 (Cantidad total demanda y abastecida: 270). Luego, para el punto de demanda 3, este sera cubierto por 150 de la planta 2 y 100 de la planta 3, por ultimo la planta 3 cubrira toda la demanda de el punto 4 (Cantidad total demanda y abastecida: 160) y el punto 5 (Cantidad total demanda y abastecida: 180). Con esto se alcanzaria el minimo costo posible que tendria que asumir la empresa en toda su operación.